### PR TITLE
feat: Add RelatedReferenceLocalesCalculator

### PR DESCRIPTION
**Pull request by ericfjosne**
This draft PR was created to check the output of the upcoming related reference locale calculator, meant to enable join operations on datasets.

To simplify the API, we remove the concept of affinityScore and replace it with an `LocaleAffinity` enum, with the following values:
- `NONE`
- `LOW`
- `HIGH`
- `SAME_OR_INTERCHANGEABLE`

With this, instead of surfacing a score and leaving implementations to figure out what a score actually means, we abstract these away behind value. This will enable us to tweak calculations under the hood if necessary, without impacting downstream users.

### CODE REVIEW

File: locales-common/src/main/java/com/spotify/i18n/locales/common/model/LocaleAffinity.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+  /** Locales are quite related, meaning they have similarities from a linguistic perspective. */
+  HIGH,
+
+  /**
+   * Locales either identify the same language, or languages that are similar to a point where a
+   * person should understand both if they understand one of them.
+   */
+  SAME_OR_INTERCHANGEABLE
+}
```
Comment (RichardNysater) [inline locales-common/src/main/java/com/spotify/i18n/locales/common/model/LocaleAffinity.java:47]: I think these explanations are good, perhaps an example could be added to each as well? 

Interestingly da-DK to no-NO has a HIGH affinity, which makes sense, but both da-DK & no-NO has NONE affinity to sv-SE which is surprising. 🤷

**Review by RichardNysater — APPROVED** [review]
Code looks good, threshold values seem reasonable based on some sample tests.

File: locales-common/src/main/java/com/spotify/i18n/locales/common/RelatedReferenceLocalesCalculator.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+ *       #getBestMatchingReferenceLocaleForTargetLocale(String)}
+ *   <li>Join datasets based on calculated reference locale only ({@link
+ *       RelatedReferenceLocale#referenceLocale()} and output of {@link
+ *       #getBestMatchingReferenceLocaleForTargetLocale(String)}), and filter based on the
+ *       calculated {@link RelatedReferenceLocale#affinity()}.
+ * </ul>
+ *
+ * @author Eric Fjøsne
+ */
+public interface RelatedReferenceLocalesCalculator {
```
Comment (RichardNysater) [inline locales-common/src/main/java/com/spotify/i18n/locales/common/RelatedReferenceLocalesCalculator.java:64]: I wonder if "NormalizedLocale" instead of "ReferenceLocale" would make the intention clearer?

Admittedly I read through the code during my morning coffee, but it wasn't until I read the above comment that it clicked that "reference" here meant the "source" definition rather than "pointing to", which made it less confusing as to how we can have a ReferenceLocale without it referring to something 😅.

That being said, the class comment explains the purpose well.

File: locales-common/src/main/java/com/spotify/i18n/locales/common/RelatedReferenceLocalesCalculator.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+      @Nullable final String languageTag);
+
+  /**
+   * For a given target locale, returns the best matching reference locale.
+   *
+   * @param languageTag target locale identifier
+   * @return the optional best matching reference locale
+   */
+  Optional<ULocale> getBestMatchingReferenceLocaleForTargetLocale(
+      @Nullable final String languageTag);
```
Comment (RichardNysater) [inline locales-common/src/main/java/com/spotify/i18n/locales/common/RelatedReferenceLocalesCalculator.java:83]: I think the language tag for both of these methods is the same, so "ForSupportedLocale" and "ForTargetLocale" can be dropped from the method names and just refer to input locale rather than supported/target.

File: locales-common/src/main/java/com/spotify/i18n/locales/common/impl/LocaleAffinityCalculatorBaseImpl.java [hunk 4]
```diff
@@ -103,8 +105,27 @@ private int getBestDistance(final String languageTag) {
...
+      return (int) ((DISTANCE_THRESHOLD - distance) / DISTANCE_THRESHOLD * 100.0);
+    }
+  }
+
+  private LocaleAffinity convertScoreToLocaleAffinity(final int score) {
+    if (score > SCORE_THRESHOLD_SAME_OR_INTERCHANGEABLE) {
+      return LocaleAffinity.SAME_OR_INTERCHANGEABLE;
+    } else if (score > SCORE_THRESHOLD_HIGH) {
+      return LocaleAffinity.HIGH;
+    } else if (score > SCORE_THRESHOLD_LOW) {
```
Comment (RichardNysater) [inline locales-common/src/main/java/com/spotify/i18n/locales/common/impl/LocaleAffinityCalculatorBaseImpl.java:121]: I noticed that I assume a threshold to be inclusive rather than exclusive unless stated otherwise, but thinking about it I'm not sure if that's actually common or just personal opinion 🤔

File: locales-common/src/main/java/com/spotify/i18n/locales/common/RelatedReferenceLocalesCalculator.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+ */
+public interface RelatedReferenceLocalesCalculator {
+
+  /**
+   * For a given supported locale, returns the list of related reference locales, along with their
+   * calculated affinities with the supported locale.
+   *
+   * @param languageTag supported locale identifier
+   * @return List of related reference locales, along with the calculated affinity
+   */
```
Comment (RichardNysater) [inline locales-common/src/main/java/com/spotify/i18n/locales/common/RelatedReferenceLocalesCalculator.java:72]: In the implementation we are sorting the result to return the highest affinity first, do we want consumers to be able to rely on that? (If so, we should document it here)

File: locales-common/src/main/java/com/spotify/i18n/locales/common/model/LocaleAffinity.java (hunk 1 above)
Comment (ericfjosne) [reply locales-common/src/main/java/com/spotify/i18n/locales/common/model/LocaleAffinity.java:47]: Agreed, but this is related to the content of the following file: https://github.com/unicode-org/cldr/blob/main/common/supplemental/languageInfo.xml 

More specifically: [da <-> no,nb](https://github.com/unicode-org/cldr/blob/main/common/supplemental/languageInfo.xml#L29-L30)

From that perspective, the output we get from this new logic is what we can expect.

We can raise an issue with Unicode on the topic of Swedish relationship to Norwegian, Bokmål and Danish.

File: locales-common/src/main/java/com/spotify/i18n/locales/common/RelatedReferenceLocalesCalculator.java (hunk 2 above)
Comment (ericfjosne) [reply locales-common/src/main/java/com/spotify/i18n/locales/common/RelatedReferenceLocalesCalculator.java:64]: This might introduce some confusion, as a the term _normalized locale_ is used more broadly. We do make use of that term in the [LanguageTagUtils class](https://github.com/spotify/java-locales/blob/main/locales-utils/src/main/java/com/spotify/i18n/locales/utils/languagetag/LanguageTagUtils.java#L43-L56) for instance.

This _reference_ type of locale is very use-case specific, to enable join operations based on locales.

The rationale for this is that we need a dedicated set of locales that can be used as _reference_ for comparisons when doing a join operation between datasets, hence the name choice.

Hope this makes sense.

File: locales-common/src/main/java/com/spotify/i18n/locales/common/RelatedReferenceLocalesCalculator.java (hunk 3 above)
Comment (ericfjosne) [reply locales-common/src/main/java/com/spotify/i18n/locales/common/RelatedReferenceLocalesCalculator.java:83]: Agreed. Removed!

File: locales-common/src/main/java/com/spotify/i18n/locales/common/model/LocaleAffinity.java (hunk 1 above)
Comment (ericfjosne) [reply locales-common/src/main/java/com/spotify/i18n/locales/common/model/LocaleAffinity.java:47]: Added a test around this specifically, for reference: https://github.com/spotify/java-locales/pull/8/commits/e8c74b188d0e0dc261b70a52d75891762c0a38a4

File: locales-common/src/main/java/com/spotify/i18n/locales/common/impl/LocaleAffinityCalculatorBaseImpl.java (hunk 4 above)
Comment (ericfjosne) [reply locales-common/src/main/java/com/spotify/i18n/locales/common/impl/LocaleAffinityCalculatorBaseImpl.java:121]: These are entirely private to the implementation anyway, it doesn't really make any difference for users whether these values are inclusive or exclusive.

File: locales-common/src/main/java/com/spotify/i18n/locales/common/RelatedReferenceLocalesCalculator.java (hunk 5 above)
Comment (ericfjosne) [reply locales-common/src/main/java/com/spotify/i18n/locales/common/RelatedReferenceLocalesCalculator.java:70]: To be honest, I believe entries in the list should be ordered by descending affinity score. But we don't have visibility on the score here, which means the list is actually sorted by enum value, which is not ideal either ...

I will see what can be done here and will respond again later on.

File: locales-common/src/test/java/com/spotify/i18n/locales/common/impl/LocaleAffinityCalculatorBaseImplTest.java [hunk 6]
```diff
@@ -82,61 +87,75 @@ void whenMatching_returnsExpectedResult(final String languageTag, final int expe
...
-        is(LocaleAffinityResult.builder().affinityScore(expectedScore).build()));
+        is(LocaleAffinityResult.builder().affinity(expectedAffinity).build()));
   }
 
   public static Stream<Arguments> whenMatching_returnsExpectedResult() {
     return Stream.of(
         // Edge cases
-        Arguments.of(" Invalid language tag ", 0),
-        Arguments.of(null, 0),
+        Arguments.of(" Invalid language tag ", NONE),
```
Comment (bridell) [inline locales-common/src/test/java/com/spotify/i18n/locales/common/impl/LocaleAffinityCalculatorBaseImplTest.java:96]: Perhaps this test should cover the edge case of "well-formed" junk input, like just the string "invalid" (which, unlike "invalid tag", is accepted by ULocale)?

Comment (ericfjosne) [reply locales-common/src/test/java/com/spotify/i18n/locales/common/impl/LocaleAffinityCalculatorBaseImplTest.java:96]: It should be covered. I will add these to make sure. Thanks for calling this out!

Comment (ericfjosne) [reply locales-common/src/test/java/com/spotify/i18n/locales/common/impl/LocaleAffinityCalculatorBaseImplTest.java:96]: Done! https://github.com/spotify/java-locales/pull/8/commits/29fddc60af44bbefc4ea9c45cef95eaaa8991f01

File: locales-common/src/main/java/com/spotify/i18n/locales/common/RelatedReferenceLocalesCalculator.java (hunk 5 above)
Comment (ericfjosne) [reply locales-common/src/main/java/com/spotify/i18n/locales/common/RelatedReferenceLocalesCalculator.java:72]: We will discard sorting for now, we can add this later on if required.

**Review by DieterVDW — APPROVED** [review]
